lib/resty/counter.lua

```diff
@@ -29,30 +32,34 @@ function _M.new(shdict_name, sync_interval)
     return nil, "shared dict \"" .. (shdict_name or "nil") .. "\" not defined"
   end
 
-  sync_interval = tonumber(sync_interval)
-  if not sync_interval or sync_interval < 0 then
-    return nil, "expect sync_interval to be a positive number"
-  end
-
   if not increments[shdict_name] then
     increments[shdict_name] = {}
   end
 
   local self = setmetatable({
     dict = ngx_shared[shdict_name],
-    sync_interval = sync_interval,
     increments = increments[shdict_name],
   }, mt)
 
-  if not timer_started[shdict_name] then
-    ngx.log(ngx.DEBUG, "start timer for shdict ", shdict_name, " on worker ", id)
-    ngx.timer.every(sync_interval, sync, self)
-    timer_started[shdict_name] = true
+  if sync_interval then
+    sync_interval = tonumber(sync_interval)
+    if not sync_interval or sync_interval < 0 then
+      return nil, "expect sync_interval to be a positive number"
+    end
```

Comment:
Minor comment, not a blocker (and not even sure if a desired change because it technically adds a compatibility breakage to the library API for not much benefit): in the Kong codebase at least, for invalid function arguments that amount to programmer error (i.e. that are not values that are derived from end-user input, which needs to be handled gracefully), we use `error(message, 2)` instead of `nil, message`) to force a crash at development time instead of letting values that would never work potentially slip by unchecked.